StreamReader.cpp

```diff
@@ -37,34 +37,58 @@ void FileStreamReader::Seek(uint64_t position) {
 	file.seekg(position, std::ios_base::beg);
 }
 
-void FileStreamReader::SeekRelative(int offset) {
+void FileStreamReader::SeekRelative(int64_t offset) {
 	file.seekg(offset, std::ios_base::cur);
 }
 
 
-MemoryStreamReader::MemoryStreamReader(char* buffer, size_t size) {
+MemoryStreamReader::MemoryStreamReader(char* buffer, uint64_t size) {
 	streamBuffer = buffer;
 	streamSize = size;
-	offset = 0;
+	position = 0;
 }
 
-void MemoryStreamReader::Read(char* buffer, size_t size) 
+void MemoryStreamReader::Read(char* buffer, uint64_t size) 
 {
-	if (offset + size > streamSize) {
+	if (position + size > streamSize) {
 		throw std::runtime_error("Size of bytes to read exceeds remaining size of buffer.");
 	}
 
-	memcpy(buffer, streamBuffer + offset, size);
-	offset += size;
+	// std::memcpy takes a _Size to copy of size_t. 
+	// Depending on the system, size_t may not be able to store the max value of uint64_t. 
+	// SIZE_MAX is the largest value that may be stored in size_t for the given system.
+	do {
+		size_t nextReadLength = SIZE_MAX;
+		// Check if remaining value of size will fit in a size_t.
+		if (size < SIZE_MAX) {
+			nextReadLength = static_cast<size_t>(size);
+		}
+
+		std::memcpy(buffer, streamBuffer + position, nextReadLength);
```

Comment:
This will overwrite the same section of buffer with each chunk.

Reply:
I didn't follow how the overwrite was occurring, but I'm switching back to the simpler model so it is a mute point.

Reply:
Yes, moot point now. As for what I meant, the destination `buffer` was not updated in the loop, only the source `streamBuffer + position`. Hence it would overwrite the same area of the destination buffer each time through the loop. The more fundamental issue though, is a buffer can never be bigger than `size_t` allows, so there's no way to iterate over `SIZE_MAX` chunks of memory and actually have it point to different memory.